examples/image_captioning/datasets.py

```diff
@@ -0,0 +1,139 @@
+_unk = 2
+_ignore = -1
+
+
+def split(sentence):
+    return sentence.lower().replace('.', ' .').replace(',', ' ,').split()
+
+
+class MsCocoDataset(dataset.DatasetMixin):
+
+    """Wraps the MSCOCO datasets and is used by the iterator to fetch data."""
+
+    def __init__(self, root_dir, data_dir, anno_file):
+        coco = COCO(os.path.join(root_dir, anno_file))
+        anns = coco.loadAnns(coco.getAnnIds())
+
+        self.coco = coco
+        self.anns = anns
+        self.vocab = None  # Later set from outside
+        self.coco_root = root_dir
+        self.coco_data = data_dir
+
+    def __len__(self):
+        return len(self.anns)
+
+    def get_example(self, i):
+        """Called by the iterator to fetch a data sample.
+
+        A data sample from MSCOCO consists of an image and its corresponding
+        caption.
+        """
+        ann = self.anns[i]
+
+        # Load the image
+        img_id = ann['image_id']
+        img_file_name = self.coco.loadImgs([img_id])[0]['file_name']
+        img = Image.open(
+            os.path.join(self.coco_root, self.coco_data, img_file_name))
+        if img.mode == 'RGB':
+            img = np.asarray(img, np.float32).transpose(2, 0, 1)
+        elif img.mode == 'L':
+            img = np.asarray(img, np.float32)
+            img = np.broadcast_to(img, (3,) + img.shape)
+        else:
+            raise ValueError('Invalid image mode {}'.format(img.mode))
+
+        # Load the caption, i.e. sequence of tokens
+        tokens = [self.vocab.get(w, _unk) for w in
+                  ['<bos>'] + split(ann['caption']) + ['<eos>']]
+        tokens = np.array(tokens, np.int32)
+
+        return img, tokens
+
+
+def get_mscoco(
+        root_dir,
+        train_dir='train2014',
+        train_anno='annotations/captions_train2014.json',
+        val_dir='val2014',
+        val_anno='annotations/captions_val2014.json',
+        unk_threshold=5):
+    """Return the training and validation datasets for MSCOCO.
+
+    The datasets can be used by the iterator during training.
+
+    A vocabulary is dynamically created based on all captions and is
+    returned as members of the training and validation dataset objects.
+    """
+    train = MsCocoDataset(root_dir, train_dir, train_anno)
+    val = MsCocoDataset(root_dir, val_dir, val_anno)
+
+    # Create a vocabulary based on the captions from both the training and the
+    # validation datasets
+    anns = train.anns + val.anns
+    captions = [ann['caption'] for ann in anns]
+
+    # Filter out rare words as UNK
+    word_counts = defaultdict(int)
+    for c in captions:
+        for w in split(c):
+            word_counts[w] += 1
+
+    # This vocabulary is needed in order to convert the words in the captions
+    # to integer tokens. When generating captions during testing, these tokens
+    # are mapped back to their corresponding words.
+    vocab = {'<bos>': _bos, '<eos>': _eos, '<unk>': _unk}
+    for w, count in six.iteritems(word_counts):
```

Comment:
Please check if this works deterministically. If not, word-to-idx dictionary `vocab` can be changed even using same dataset.
The current implementation of `predict.py` uses a vocabulary which is re-constructed by reading dataset. So, if not deterministic, vocabulary used in the `predict.py` can be different from a vocabulary used during training a loaded model.

For example, I think the processing below is deterministic.

```
for w, count in sorted(word_counts.items(), lambda x:(-x[1], x[0])):
```

Reply:
I think `sorted(word_counts.items())` is sufficient to sort the vocab alphanumerically. Updated the code to do so.

Reply:
OK!